### PR TITLE
[EXE-2119] Implement Write to Transient Table

### PR DESCRIPTION
**Pull request by yannistze**
Adding Support for Write to Transient Table Snowflake

## Test Notes
```
-> % sbt compile
...
[success] Total time: 0 s, completed Feb 15, 2024, 1:35:45 AM
```

```
export SNOWFLAKE_TEST_ACCOUNT="aws"
export SKIP_BIG_DATA_TEST=true
export SPARK_LOCAL_IP=127.0.0.1
export SPARK_CONN_ENV_SFURL=xxxx
export SPARK_CONN_ENV_SFUSER=xxx
export SPARK_CONN_ENV_SFPASSWORD=xxx
export SPARK_CONN_ENV_SFWAREHOUSE=DEMO_WH
export SPARK_CONN_ENV_SFDATABASE=TEST_DB
export SPARK_CONN_ENV_SFSCHEMA=PUBLIC
export SPARK_CONN_ENV_DBTABLE=SPARK_SNOWFLAKE_CI
```

```
sbt "it:testOnly net.snowflake.spark.snowflake.JDBCSuite" >> /Users/yannis.tzemos/Documents/CodeSnips/snow_test_output_jdbcsuite.txt
...
[info] JDBCSuite:
[info] - create and drop table
[info] - create and drop stage
[info] - create and drop pipe
[info] - AIQ create and drop transient table
[info] - test schema
+---+---+
|NUM|STR|
+---+---+
+---+---+

[info] - copy from query
[info] - copy from query using udf
[info] - test union
[info] Run completed in 16 seconds, 255 milliseconds.
[info] Total number of tests run: 8
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 8, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 17 s, completed Feb 15, 2024, 1:36:15 AM
```

- Published `"2.11.3-aiq11-yt2"` of the Snowflake Connector
- Picked up the Snowflake Connector version in Flame (locally)

```
~/flame3.3/aiq/maven-clean.sh"
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  6.407 s
[INFO] Finished at: 2024-02-15T01:38:36-05:00
[INFO] ------------------------------------------------------------------------
```

```
~/flame3.3/aiq/maven-compile.sh
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  08:50 min
[INFO] Finished at: 2024-02-15T01:48:13-05:00
[INFO] ------------------------------------------------------------------------
```

```
~/flame3.3/aiq/maven-install.sh
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  12:42 min
[INFO] Finished at: 2024-02-15T02:01:37-05:00
[INFO] ------------------------------------------------------------------------
```

```
~/flame3.3/aiq/spark-shell.sh
...
scala> val df = spark.read.format(snowflakeRelationClass).options(sfOptions).option("dbtable", "dim_customer").load().select(""""customer_id"""")
df: org.apache.spark.sql.DataFrame = ["customer_id": bigint]
```

```
scala> df.alias("df1").join(df.alias("df2"), Seq(""""customer_id""""), "left_outer").select(col("""df1.`"customer_id"`""").alias("customer_id_1"), col("""df2.`"customer_id"`""").alias("customer_id_2")).write.format(snowflakeRelationClass).options(sfOptions).option("dbtable", "yannis_test_table").mode("overwrite").save()
```

Screenshot 2024-02-15 at 02 34 39

Screenshot 2024-02-15 at 02 38 08

Screenshot 2024-02-15 at 02 35 02

```
scala> df.alias("df1").join(df.alias("df2"), Seq(""""customer_id""""), "left_outer").select(col("""df1.`"customer_id"`""").alias("customer_id_1"), col("""df2.`"customer_id"`""").alias("customer_id_2")).write.format(snowflakeRelationClass).options(sfOptions).option("dbtable", "yannis_test_table").option("write_to_transient_table", "true").mode("overwrite").save()
```

Screenshot 2024-02-15 at 02 39 27

Screenshot 2024-02-15 at 02 41 25

Screenshot 2024-02-15 at 02 41 48

## Deploy
- Merge and `sbt publish`
- Pick up new Connector version in Flame
- Deploy new Clusters

### CODE REVIEW

**Review by dorisZ017 — APPROVED** [review]

File: src/it/scala/net/snowflake/spark/snowflake/JDBCSuite.scala [hunk 1]
```diff
@@ -103,6 +103,46 @@ class JDBCSuite extends IntegrationSuiteBase {
...
+
+    val transientParams = params.copy(
+      parameters = params.parameters.map {
+        case ("write_to_transient_table", "false") => "write_to_transient_table" -> "true"
+        case kv => kv
+      }
+    )
+
+    def validate(expectedKind: String): Unit = {
+      assert(conn.tableExists(name))
```
Comment (MasterDDT) [inline src/it/scala/net/snowflake/spark/snowflake/JDBCSuite.scala:120]: Why so many extra newlines

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala [hunk 2]
```diff
@@ -380,7 +380,8 @@ private[snowflake] object DefaultJDBCWrapper extends JDBCWrapper {
                     bindVariableEnabled: Boolean = true): Unit =
       (ConstantString("create") +
         (if (overwrite) "or replace" else "") +
-        (if (temporary) "temporary" else "") + "table" +
+        (if (temporary) "temporary" else "") +
+        (if (!temporary && params.writeToTransientTableCheck) "transient" else "") + "table" +
```
Comment (MasterDDT) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:384]: Nice, this is the only change needed?

File: src/it/scala/net/snowflake/spark/snowflake/JDBCSuite.scala [hunk 3]
```diff
@@ -103,6 +103,46 @@ class JDBCSuite extends IntegrationSuiteBase {
...
   }
 
+  test("AIQ create and drop transient table") {
+
+    val name = s"spark_test_table_$randomSuffix"
+    val schema =
+      new StructType(Array(StructField("num", IntegerType, nullable = false)))
+
+    val transientParams = params.copy(
+      parameters = params.parameters.map {
```
Comment (MasterDDT) [inline src/it/scala/net/snowflake/spark/snowflake/JDBCSuite.scala:113]: Why not just do `params.parameters.set(write_to_transient_table, true)`

File: src/it/scala/net/snowflake/spark/snowflake/JDBCSuite.scala (hunk 1 above)
Comment (yannistze) [reply src/it/scala/net/snowflake/spark/snowflake/JDBCSuite.scala:120]: Just to make sure it's fully tested with all the flag combinations for `overwrite`, `temporary` 🙏

File: src/it/scala/net/snowflake/spark/snowflake/IntegrationSuiteBase.scala [hunk 4]
```diff
@@ -179,7 +180,25 @@ trait IntegrationSuiteBase
     }
   }
 
+  // Utility function to test the `kind` of the provided table
+  def testTableKind(testTableName: String, result: String): Unit = {
```
Comment (MasterDDT) [inline src/it/scala/net/snowflake/spark/snowflake/IntegrationSuiteBase.scala:184]: Interesting, i guess you dont need any special markings on this method...it just gets picked up as a new test?

**Review by MasterDDT — APPROVED** [review]

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala (hunk 2 above)
Comment (yannistze) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:384]: Yeap, only change needed 🙏 Managed to test it with Flame as well to be certain 🙏

File: src/it/scala/net/snowflake/spark/snowflake/JDBCSuite.scala (hunk 3 above)
Comment (yannistze) [reply src/it/scala/net/snowflake/spark/snowflake/JDBCSuite.scala:113]: @MasterDDT, actually there is no `set` function for that `map` 🤔

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala (hunk 2 above)
Comment (MasterDDT) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:384]: Also i guess if i do temporary=true and transient=true...it becomes temporary only?

Comment (yannistze) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:384]: Yeap and this is covered/tested by the Integration test as well 👍

File: src/it/scala/net/snowflake/spark/snowflake/JDBCSuite.scala (hunk 3 above)
Comment (yannistze) [reply src/it/scala/net/snowflake/spark/snowflake/JDBCSuite.scala:113]: Scratch the above, there is `updated` 👍

**Pull request merged**